### PR TITLE
Add ThreadPool to Host Workspace

### PR DESCRIPTION
**Pull request by klecki**
Fill the pointer in workspace policies

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/pipeline/executor/workspace_policy.h [hunk 1]
```diff
@@ -105,6 +105,18 @@ void SetupInputOutput(op_type_to_workspace_t<op_type> &ws, const OpGraph &graph,
   }
 }
 
+template <OpType op_type>
+inline void SetupThreadPool(op_type_to_workspace_t<op_type> &ws, const OpGraph &,
+                           const OpNode &, ThreadPool *, const QueueIdxs) {
```
Comment (jantonguirao) [inline dali/pipeline/executor/workspace_policy.h]: align

File: dali/pipeline/executor/workspace_policy.h [hunk 2]
```diff
@@ -105,6 +105,18 @@ void SetupInputOutput(op_type_to_workspace_t<op_type> &ws, const OpGraph &graph,
...
+  /* No-op if we are not CPU */
+}
+
+template <>
+inline void SetupThreadPool<OpType::CPU>(op_type_to_workspace_t<OpType::CPU> &ws, const OpGraph &,
+                           const OpNode &, ThreadPool *thread_pool, const QueueIdxs) {
```
Comment (jantonguirao) [inline dali/pipeline/executor/workspace_policy.h]: align

File: dali/pipeline/executor/workspace_policy.h [hunk 3]
```diff
@@ -152,12 +164,12 @@ inline void SetupStreamsAndEvents<OpType::GPU>(
 template <OpType op_type>
 op_type_to_workspace_t<op_type> CreateWorkspace(
     const OpGraph &graph, const OpNode &node,
-    const std::vector<tensor_data_store_queue_t> &tensor_to_store_queue,
+    const std::vector<tensor_data_store_queue_t> &tensor_to_store_queue, ThreadPool *thread_pool,
```
Comment (jantonguirao) [inline dali/pipeline/executor/workspace_policy.h]: weird line length

File: dali/pipeline/executor/workspace_policy.h (hunk 1 above)
Comment (klecki) [reply dali/pipeline/executor/workspace_policy.h]: Done

File: dali/pipeline/executor/workspace_policy.h (hunk 2 above)
Comment (klecki) [reply dali/pipeline/executor/workspace_policy.h]: Done

File: dali/pipeline/executor/workspace_policy.h (hunk 3 above)
Comment (klecki) [reply dali/pipeline/executor/workspace_policy.h]: That's how clang-format did it. The template parameters are looong, and it's hard to fit them. I can split it to separate lines, but that's not better by much.

**Comment by klecki** [comment]
!build

File: dali/pipeline/executor/workspace_policy.h [hunk 4]
```diff
@@ -390,14 +408,15 @@ struct AOT_WS_Policy<SeparateQueuePolicy> {
   template <OpType op_type, OpType group_as = op_type, typename T>
   void PlaceWorkspace(T &workspaces, QueueIdxs idxs, const OpGraph &graph,
                       const std::vector<tensor_data_store_queue_t> &tensor_to_store_queue,
-                      cudaStream_t mixed_op_stream, cudaStream_t gpu_op_stream,
+                      ThreadPool *thread_pool, cudaStream_t mixed_op_stream,
```
Comment (mzient) [inline dali/pipeline/executor/workspace_policy.h]: Move each argument to its own line or group them in a meaningful way.

File: dali/pipeline/executor/workspace_policy.h [hunk 5]
```diff
@@ -390,14 +408,15 @@ struct AOT_WS_Policy<SeparateQueuePolicy> {
...
+                      cudaStream_t gpu_op_stream,
                       const std::vector<std::vector<cudaEvent_t>> &mixed_op_events) {
     int sequential_ws_idx = SequentialIndex(idxs, depths_, group_as);
     workspaces[sequential_ws_idx].resize(graph.NumOp(op_type));
     for (OpPartitionId partition_idx = 0; partition_idx < graph.NumOp(op_type); partition_idx++) {
       auto &node = graph.Node(op_type, partition_idx);
       workspaces[sequential_ws_idx][partition_idx] =
-          CreateWorkspace<op_type>(graph, node, tensor_to_store_queue, mixed_op_stream,
+          CreateWorkspace<op_type>(graph, node, tensor_to_store_queue, thread_pool, mixed_op_stream,
```
Comment (mzient) [inline dali/pipeline/executor/workspace_policy.h]: See above: I'd rather see the two streams next to each other.

**Comment by klecki** [comment]
!build

File: dali/pipeline/executor/workspace_policy.h (hunk 4 above)
Comment (klecki) [reply dali/pipeline/executor/workspace_policy.h]: done

File: dali/pipeline/executor/workspace_policy.h (hunk 5 above)
Comment (klecki) [reply dali/pipeline/executor/workspace_policy.h]: done

**Pull request merged**